sparse_strips/vello_common/src/filter_effects.rs

```diff
@@ -0,0 +1,1142 @@
+    ///
+    /// Creates a simple filter graph with a single primitive.
+    /// Use this for direct access to low-level SVG filter operations.
+    pub fn from_primitive(primitive: FilterPrimitive) -> Self {
+        let mut graph = FilterGraph::new();
+        let filter_id = graph.add(primitive, None);
+        graph.set_output(filter_id);
+
+        Self {
+            graph,
+            bounds: None,
+        }
+    }
+
+    /// Calculate the total bounds expansion for this filter in pixels.
+    ///
+    /// Returns how many extra pixels are needed around the filtered region
+    /// to correctly compute the filter effect. For example, a blur filter
+    /// needs to sample beyond the original bounds to avoid edge artifacts.
+    ///
+    /// This value is cached and computed incrementally as primitives are added
+    /// to the filter graph.
+    pub fn bounds_expansion(&self) -> BoundsExpansion {
+        self.graph.bounds_expansion()
+    }
+}
+
+/// Describes how many pixels a filter expands the required processing region
+/// in each direction.
+///
+/// Filters like blur need to sample pixels beyond the original bounds to produce
+/// correct results. This struct describes the padding needed in each direction.
+/// All values are non-negative.
+#[derive(Debug, Clone, Copy, PartialEq)]
+pub struct BoundsExpansion {
+    /// Extra pixels needed to the left of the original bounds.
+    pub left: u16,
+    /// Extra pixels needed above the original bounds.
+    pub top: u16,
+    /// Extra pixels needed to the right of the original bounds.
+    pub right: u16,
+    /// Extra pixels needed below the original bounds.
+    pub bottom: u16,
+}
+
+impl BoundsExpansion {
+    /// Create a zero expansion (no change).
+    pub const fn zero() -> Self {
+        Self {
+            left: 0,
+            top: 0,
+            right: 0,
+            bottom: 0,
+        }
+    }
+
+    /// Create a uniform expansion in all directions.
+    pub fn uniform(radius: u16) -> Self {
+        Self {
+            left: radius,
+            top: radius,
+            right: radius,
+            bottom: radius,
+        }
+    }
+
+    /// Combine two expansions by taking the maximum in each direction.
+    /// Used when composing multiple filters.
+    pub fn max(&self, other: &Self) -> Self {
+        Self {
+            left: self.left.max(other.left),
+            top: self.top.max(other.top),
+            right: self.right.max(other.right),
+            bottom: self.bottom.max(other.bottom),
+        }
+    }
+}
+
+/// A directed acyclic graph (DAG) of filter operations.
+///
+/// The graph represents a pipeline of filter primitives where outputs of some
+/// primitives can be used as inputs to others. Each primitive has a unique `FilterId`.
+#[derive(Debug, Clone, PartialEq)]
+pub struct FilterGraph {
+    /// All filter primitives in the graph, stored in insertion order.
+    pub primitives: Vec<FilterPrimitive>,
```

Comment:
We could probably make this a `SmallVec` of 1 or 2 size?

Reply:
Yes, good idea! I think that would work better for the common case.

Fixed.